File: frontend/src/components/benchmark/content/leiden/implementations/leiden.go
package main

import (
    "math/rand"
    "time"
)

type Leiden struct {
    adjacency  map[int]map[int]bool
    nodes      map[int]bool
    communities map[int]int
    weights    map[[2]int]float64
    m          float64
}

func NewLeiden() *Leiden {
    return &Leiden{
        adjacency:  make(map[int]map[int]bool),
        nodes:      make(map[int]bool),
        communities: make(map[int]int),
        weights:    make(map[[2]int]float64),
        m:          0.0,
    }
}

func (l *Leiden) AddEdge(u, v int, weight float64) {
    if l.adjacency[u] == nil {
        l.adjacency[u] = make(map[int]bool)
    }
    if l.adjacency[v] == nil {
        l.adjacency[v] = make(map[int]bool)
    }

    l.adjacency[u][v] = true
    l.adjacency[v][u] = true

    l.nodes[u] = true
    l.nodes[v] = true

    if u > v {
        u, v = v, u
    }
    l.weights[[2]int{u, v}] += weight
    l.m += weight
}

func (l *Leiden) BuildFromEdges(edges [][2]int) {
    for _, edge := range edges {
        l.AddEdge(edge[0], edge[1], 1.0)
    }
}

func (l *Leiden) degree(node int) float64 {
    deg := 0.0
    for neighbor := range l.adjacency[node] {
        u, v := node, neighbor
        if u > v {
            u, v = v, u
        }
        deg += l.weights[[2]int{u, v}]
    }
    return deg
}

func (l *Leiden) weight(u, v int) float64 {
    if u > v {
        u, v = v, u
    }
    return l.weights[[2]int{u, v}]
}

func (l *Leiden) communityDegree(node, community int) float64 {
    deg := 0.0
    for neighbor := range l.adjacency[node] {
        if l.communities[neighbor] == community {
            deg += l.weight(node, neighbor)
        }
    }
    return deg
}

func (l *Leiden) totalDegree(community int) float64 {
    total := 0.0
    for node := range l.nodes {
        if l.communities[node] == community {
            total += l.degree(node)
        }
    }
    return total
}

func (l *Leiden) modularityGain(node, community int) float64 {
    ki := l.degree(node)
    kic := l.communityDegree(node, community)
    total := l.totalDegree(community)
    return (kic - (total*ki)/(2*l.m)) / l.m
}

func (l *Leiden) initializeCommunities() {
    for node := range l.nodes {
        l.communities[node] = node
    }
}

func (l *Leiden) refinePartition() bool {
    changed := false
    nodeList := make([]int, 0, len(l.nodes))
    for node := range l.nodes {
        nodeList = append(nodeList, node)
    }

    rand.Seed(time.Now().UnixNano())
    rand.Shuffle(len(nodeList), func(i, j int) {
        nodeList[i], nodeList[j] = nodeList[j], nodeList[i]
    })

    for _, node := range nodeList {
        currentCommunity := l.communities[node]
        bestCommunity := currentCommunity
        bestGain := 0.0

        communitiesSeen := make(map[int]bool)
        for neighbor := range l.adjacency[node] {
            community := l.communities[neighbor]
            if communitiesSeen[community] {
                continue
            }
            communitiesSeen[community] = true

            if community == currentCommunity {
                continue
            }

            gain := l.modularityGain(node, community)
            if gain > bestGain {
                bestGain = gain
                bestCommunity = community
            }
        }

        if bestCommunity != currentCommunity {
            l.communities[node] = bestCommunity
            changed = true
        }
    }

    return changed
}

func (l *Leiden) aggregateNetwork() {
    newAdjacency := make(map[int]map[int]bool)
    newNodes := make(map[int]bool)
    newWeights := make(map[[2]int]float64)
    newM := 0.0

    communityMap := make(map[int]int)
    nextId := 0
    communitySet := make(map[int]bool)
    for _, community := range l.communities {
        if !communitySet[community] {
            communitySet[community] = true
            communityMap[community] = nextId
            nextId++
        }
    }

    for node, community := range l.communities {
        newCommunity := communityMap[community]
        if newAdjacency[newCommunity] == nil {
            newAdjacency[newCommunity] = make(map[int]bool)
        }
        newNodes[newCommunity] = true
    }

    for edge, weight := range l.weights {
        u := edge[0]
        v := edge[1]
        cu := communityMap[l.communities[u]]
        cv := communityMap[l.communities[v]]

        if cu == cv {
            newM += weight
            continue
        }

        if cu > cv {
            cu, cv = cv, cu
        }
        newWeights[[2]int{cu, cv}] += weight
        if newAdjacency[cu] == nil {
            newAdjacency[cu] = make(map[int]bool)
        }
        if newAdjacency[cv] == nil {
            newAdjacency[cv] = make(map[int]bool)
        }
        newAdjacency[cu][cv] = true
        newAdjacency[cv][cu] = true
        newM += weight
    }

    for community := range newNodes {
        if newAdjacency[community] == nil {
            newAdjacency[community] = make(map[int]bool)
        }
    }

    l.adjacency = newAdjacency
    l.nodes = newNodes
    l.weights = newWeights
    l.m = newM

    newCommunities := make(map[int]int)
    for node, community := range l.communities {
        newCommunities[node] = communityMap[community]
    }
    l.communities = newCommunities
}

func (l *Leiden) fastLocalMove() bool {
    changed := false
    nodeList := make([]int, 0, len(l.nodes))
    for node := range l.nodes {
        nodeList = append(nodeList, node)
    }

    rand.Seed(time.Now().UnixNano())
    rand.Shuffle(len(nodeList), func(i, j int) {
        nodeList[i], nodeList[j] = nodeList[j], nodeList[i]
    })

    for _, node := range nodeList {
        bestCommunity := l.communities[node]
        bestGain := 0.0

        for neighbor := range l.adjacency[node] {
            community := l.communities[neighbor]
            if community == l.communities[node] {
                continue
            }
            gain := l.modularityGain(node, community)
            if gain > bestGain {
                bestGain = gain
                bestCommunity = community
            }
        }

        if bestCommunity != l.communities[node] {
            l.communities[node] = bestCommunity
            changed = true
        }
    }

    return changed
}

func (l *Leiden) DetectCommunities() map[int]int {
    l.initializeCommunities()

    for {
        improved := false

        for l.refinePartition() {
            improved = true
        }

        if !improved {
            break
        }

        for l.fastLocalMove() {
            improved = true
        }

        if !improved {
            break
        }

        l.aggregateNetwork()
    }

    return l.communities
}

File: frontend/src/components/benchmark/content/leiden/implementations/leiden.js
class Leiden {
    constructor() {
        this.adjacency = new Map();
        this.nodes = new Set();
        this.communities = new Map();
        this.weights = new Map();
        this.m = 0.0;
    }

    addEdge(u, v, weight = 1.0) {
        if (!this.adjacency.has(u)) {
            this.adjacency.set(u, new Set());
        }
        if (!this.adjacency.has(v)) {
            this.adjacency.set(v, new Set());
        }

        this.adjacency.get(u).add(v);
        this.adjacency.get(v).add(u);

        this.nodes.add(u);
        this.nodes.add(v);

        const key = u < v ? `${u},${v}` : `${v},${u}`;
        this.weights.set(key, (this.weights.get(key) || 0) + weight);
        this.m += weight;
    }

    buildFromEdges(edges) {
        for (const [u, v] of edges) {
            this.addEdge(u, v);
        }
    }

    degree(node) {
        let deg = 0.0;
        for (const neighbor of this.adjacency.get(node)) {
            deg += this.weight(node, neighbor);
        }
        return deg;
    }

    weight(u, v) {
        const key = u < v ? `${u},${v}` : `${v},${u}`;
        return this.weights.get(key) || 0;
    }

    communityDegree(node, community) {
        let deg = 0.0;
        for (const neighbor of this.adjacency.get(node)) {
            if (this.communities.get(neighbor) === community) {
                deg += this.weight(node, neighbor);
            }
        }
        return deg;
    }

    totalDegree(community) {
        let total = 0.0;
        for (const node of this.nodes) {
            if (this.communities.get(node) === community) {
                total += this.degree(node);
            }
        }
        return total;
    }

    modularityGain(node, community) {
        const ki = this.degree(node);
        const kic = this.communityDegree(node, community);
        const total = this.totalDegree(community);
        return (kic - (total * ki) / (2 * this.m)) / this.m;
    }

    initializeCommunities() {
        for (const node of this.nodes) {
            this.communities.set(node, node);
        }
    }

    refinePartition() {
        let changed = false;
        const nodeList = Array.from(this.nodes);
        this.shuffle(nodeList);

        for (const node of nodeList) {
            const currentCommunity = this.communities.get(node);
            let bestCommunity = currentCommunity;
            let bestGain = 0.0;

            const communitiesSeen = new Set();
            for (const neighbor of this.adjacency.get(node)) {
                const community = this.communities.get(neighbor);
                if (communitiesSeen.has(community)) continue;
                communitiesSeen.add(community);

                if (community === currentCommunity) continue;

                const gain = this.modularityGain(node, community);
                if (gain > bestGain) {
                    bestGain = gain;
                    bestCommunity = community;
                }
            }

            if (bestCommunity !== currentCommunity) {
                this.communities.set(node, bestCommunity);
                changed = true;
            }
        }

        return changed;
    }

    aggregateNetwork() {
        const newAdjacency = new Map();
        const newNodes = new Set();
        const newWeights = new Map();
        let newM = 0.0;

        const communityMap = new Map();
        let nextId = 0;
        const communitySet = new Set(this.communities.values());
        for (const community of communitySet) {
            communityMap.set(community, nextId++);
        }

        for (const [node, community] of this.communities) {
            const newCommunity = communityMap.get(community);
            if (!newAdjacency.has(newCommunity)) {
                newAdjacency.set(newCommunity, new Set());
            }
            newNodes.add(newCommunity);
        }

        for (const [key, weight] of this.weights) {
            const [u, v] = key.split(',').map(Number);
            const cu = communityMap.get(this.communities.get(u));
            const cv = communityMap.get(this.communities.get(v));

            if (cu === cv) {
                newM += weight;
                continue;
            }

            const newKey = cu < cv ? `${cu},${cv}` : `${cv},${cu}`;
            newWeights.set(newKey, (newWeights.get(newKey) || 0) + weight);
            
            if (!newAdjacency.has(cu)) newAdjacency.set(cu, new Set());
            if (!newAdjacency.has(cv)) newAdjacency.set(cv, new Set());
            
            newAdjacency.get(cu).add(cv);
            newAdjacency.get(cv).add(cu);
            newM += weight;
        }

        for (const community of newNodes) {
            if (!newAdjacency.has(community)) {
                newAdjacency.set(community, new Set());
            }
        }

        this.adjacency = newAdjacency;
        this.nodes = newNodes;
        this.weights = newWeights;
        this.m = newM;

        const newCommunities = new Map();
        for (const [node, community] of this.communities) {
            newCommunities.set(node, communityMap.get(community));
        }
        this.communities = newCommunities;
    }

    fastLocalMove() {
        let changed = false;
        const nodeList = Array.from(this.nodes);
        this.shuffle(nodeList);

        for (const node of nodeList) {
            let bestCommunity = this.communities.get(node);
            let bestGain = 0.0;

            for (const neighbor of this.adjacency.get(node)) {
                const community = this.communities.get(neighbor);
                if (community === this.communities.get(node)) continue;
                const gain = this.modularityGain(node, community);
                if (gain > bestGain) {
                    bestGain = gain;
                    bestCommunity = community;
                }
            }

            if (bestCommunity !== this.communities.get(node)) {
                this.communities.set(node, bestCommunity);
                changed = true;
            }
        }

        return changed;
    }

    shuffle(array) {
        for (let i = array.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [array[i], array[j]] = [array[j], array[i]];
        }
        return array;
    }

    detectCommunities() {
        this.initializeCommunities();

        while (true) {
            let improved = false;

            while (this.refinePartition()) {
                improved = true;
            }

            if (!improved) break;

            while (this.fastLocalMove()) {
                improved = true;
            }

            if (!improved) break;

            this.aggregateNetwork();
        }

        return this.communities;
    }
}

module.exports = Leiden;